diff --git a/src/ec_mult.h b/src/ec_mult.h
index eb9c874..3bcce34 100644
--- a/src/ec_mult.h
+++ b/src/ec_mult.h
@@ -79,24 +79,23 @@ static void ptype##_gather_booth_w5(ptype *restrict p, const ptype table[16], \
 \
 static void ptype##_precompute(ptype *row, const ptype *point) \
 { \
-    row--;  /* row[-1] is implicit infinity */\
 \
-    vec_copy(row + 1, point, sizeof(ptype));    /* row[ 1]=p*1     */\
-    ptype##_double(row + 2,  point);            /* row[ 2]=p*(1+1) */\
-    ptype##_add   (row + 3,  row + 2, row + 1); /* row[ 3]=p*(2+1) */\
-    ptype##_double(row + 4,  row + 2);          /* row[ 4]=p*(2+2) */\
-    ptype##_add   (row + 5,  row + 3, row + 2); /* row[ 5]=p*(3+2) */\
-    ptype##_double(row + 6,  row + 3);          /* row[ 6]=p*(3+3) */\
-    ptype##_add   (row + 7,  row + 4, row + 3); /* row[ 7]=p*(4+3) */\
-    ptype##_double(row + 8,  row + 4);          /* row[ 8]=p*(4+4) */\
-    ptype##_add   (row + 9,  row + 5, row + 4); /* row[ 9]=p*(5+4) */\
-    ptype##_double(row + 10, row + 5);          /* row[10]=p*(5+5) */\
-    ptype##_add   (row + 11, row + 6, row + 5); /* row[11]=p*(6+5) */\
-    ptype##_double(row + 12, row + 6);          /* row[12]=p*(6+6) */\
-    ptype##_add   (row + 13, row + 7, row + 6); /* row[13]=p*(7+6) */\
-    ptype##_double(row + 14, row + 7);          /* row[14]=p*(7+7) */\
-    ptype##_add   (row + 15, row + 8, row + 7); /* row[15]=p*(8+7) */\
-    ptype##_double(row + 16, row + 8);          /* row[16]=p*(8+8) */\
+    vec_copy(row + 1-1, point, sizeof(ptype));    /* row[ 1]=p*1     */\
+    ptype##_double(row + 2-1,  point);            /* row[ 2]=p*(1+1) */\
+    ptype##_add   (row + 3-1,  row + 2-1, row + 1-1); /* row[ 3]=p*(2+1) */\
+    ptype##_double(row + 4-1,  row + 2-1);          /* row[ 4]=p*(2+2) */\
+    ptype##_add   (row + 5-1,  row + 3-1, row + 2-1); /* row[ 5]=p*(3+2) */\
+    ptype##_double(row + 6-1,  row + 3-1);          /* row[ 6]=p*(3+3) */\
+    ptype##_add   (row + 7-1,  row + 4-1, row + 3-1); /* row[ 7]=p*(4+3) */\
+    ptype##_double(row + 8-1,  row + 4-1);          /* row[ 8]=p*(4+4) */\
+    ptype##_add   (row + 9-1,  row + 5-1, row + 4-1); /* row[ 9]=p*(5+4) */\
+    ptype##_double(row + 10-1, row + 5-1);          /* row[10]=p*(5+5) */\
+    ptype##_add   (row + 11-1, row + 6-1, row + 5-1); /* row[11]=p*(6+5) */\
+    ptype##_double(row + 12-1, row + 6-1);          /* row[12]=p*(6+6) */\
+    ptype##_add   (row + 13-1, row + 7-1, row + 6-1); /* row[13]=p*(7+6) */\
+    ptype##_double(row + 14-1, row + 7-1);          /* row[14]=p*(7+7) */\
+    ptype##_add   (row + 15-1, row + 8-1, row + 7-1); /* row[15]=p*(8+7) */\
+    ptype##_double(row + 16-1, row + 8-1);          /* row[16]=p*(8+8) */\
 } \
 \
 static void ptype##s_mult_w5(ptype *ret, \
