diff --git a/src/sha256.h b/src/sha256.h
index 0e10703..8b35ae4 100644
--- a/src/sha256.h
+++ b/src/sha256.h
@@ -20,9 +20,13 @@
 #define sha256_bcopy blst_sha256_bcopy
 #define sha256_emit  blst_sha256_emit
 
+#include "string.h"
 void sha256_block_data_order(unsigned int *h, const void *inp, size_t blocks);
 void sha256_hcopy(unsigned int dst[8], const unsigned int src[8]);
-void sha256_bcopy(void *dst, const void *src, size_t len);
+void sha256_bcopy(void *dst, const void *src, size_t len)
+{
+  memcpy(dst, src, len);
+};
 
 /*
  * If SHA256_CTX conflicts with something, just redefine it to alternative
